File: 2015/Day2/src/day2_ai.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <algorithm>

struct Present {
    int length, width, height;
};

// Calculate wrapping paper needed for one present
int calculateWrappingPaper(const Present& p) {
    int side1 = p.length * p.width;
    int side2 = p.width * p.height;
    int side3 = p.height * p.length;
    int surfaceArea = 2 * side1 + 2 * side2 + 2 * side3;
    int slack = std::min({side1, side2, side3});
    return surfaceArea + slack;
}

// Calculate ribbon needed for one present
int calculateRibbon(const Present& p) {
    // Smallest perimeter of any face
    int perimeter1 = 2 * (p.length + p.width);
    int perimeter2 = 2 * (p.width + p.height);
    int perimeter3 = 2 * (p.height + p.length);
    int wrap = std::min({perimeter1, perimeter2, perimeter3});
    // Bow is the volume of the present
    int bow = p.length * p.width * p.height;
    return wrap + bow;
}

int main() {
    std::ifstream file("input.txt"); // Assume input is in a file named input.txt
    std::string line;
    std::vector<Present> presents;
    
    // Read input
    while (std::getline(file, line)) {
        int l, w, h;
        char x; // To capture 'x' characters
        std::stringstream ss(line);
        ss >> l >> x >> w >> x >> h;
        presents.push_back({l, w, h});
    }
    
    // Part 1: Calculate total wrapping paper
    long long totalPaper = 0;
    for (const auto& p : presents) {
        totalPaper += calculateWrappingPaper(p);
    }
    
    // Part 2: Calculate total ribbon
    long long totalRibbon = 0;
    for (const auto& p : presents) {
        totalRibbon += calculateRibbon(p);
    }
    
    // Output results
    std::cout << "Total wrapping paper needed: " << totalPaper << " square feet\n";
    std::cout << "Total ribbon needed: " << totalRibbon << " feet\n";
    
    return 0;
}

File: 2015/Day2/src/day2_ai.py
def calculate_wrapping_and_ribbon(input_file):
    total_paper = 0
    total_ribbon = 0
    
    # Read input file
    with open(input_file, 'r') as file:
        for line in file:
            # Parse dimensions
            l, w, h = map(int, line.strip().split('x'))
            
            # Part 1: Wr colère

            # Surface area = 2*l*w + 2*w*h + 2*h*l
            surface_area = 2 * l * w + 2 * w * h + 2 * h * l
            # Smallest side area for slack
            sides = [l * w, w * h, h * l]
            slack = min(sides)
            # Total paper for this present
            total_paper += surface_area + slack
            
            # Part 2: Ribbon
            # Smallest perimeter of any face
            perimeters = [2 * (l + w), 2 * (w + h), 2 * (h + l)]
            ribbon_wrap = min(perimeters)
            # Volume for bow
            bow = l * w * h
            # Total ribbon for this present
            total_ribbon += ribbon_wrap + bow
    
    return total_paper, total_ribbon

# Assume input is in a file named 'input.txt'
input_file = 'input.txt'
paper, ribbon = calculate_wrapping_and_ribbon(input_file)
print(f"Total wrapping paper needed: {paper} square feet")
print(f"Total ribbon needed: {ribbon} feet")

# Verify against provided answers
assert paper == 1606483, f"Expected 1606483 square feet of paper, got {paper}"
assert ribbon == 3842356, f"Expected 3842356 feet of ribbon, got {ribbon}"